lib/Runtime/Language/SourceTextModuleRecord.cpp

```diff
@@ -765,7 +844,52 @@ namespace Js
         CleanupBeforeExecution();
 
         Arguments outArgs(CallInfo(CallFlags_Value, 0), nullptr);
-        return rootFunction->CallRootFunction(outArgs, scriptContext, true);
+
+        Var ret = nullptr;
+        try
+        {
+            AUTO_NESTED_HANDLED_EXCEPTION_TYPE((ExceptionType)(ExceptionType_OutOfMemory | ExceptionType_JavascriptException));
+            ENTER_SCRIPT_IF(scriptContext, true, false, false, !scriptContext->GetThreadContext()->IsScriptActive(),
+            {
+                ret = rootFunction->CallRootFunction(outArgs, scriptContext, true);
+            });
+        }
+        catch (const Js::JavascriptException &err)
+        {
+            if (this->promise != nullptr)
+            {
+                Var errorObject = err.GetAndClear()->GetThrownObject(scriptContext);
+                AssertMsg(errorObject != nullptr, "ModuleEvaluation: null error object thrown from root function");
+                if (errorObject != nullptr)
+                {
+                    this->errorObject = errorObject;
+                    ResolveOrRejectDynamicImportPromise(false, errorObject, scriptContext, this);
+                    return scriptContext->GetLibrary()->GetUndefined();
+                }
+                else
+                {
+                    Throw::InternalError();
+                }
+            }
+
+            throw(err);
```

Comment:
throw outside of catch handler. (look at other example in the code base, where they use different function for re-throw).